source/opt/cfg_cleanup_pass.cpp

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a pass to cleanup the CFG to remove superfluous
+// constructs (e.g., unreachable basic blocks, empty control flow structures,
+// etc)
+
+#include <queue>
+#include <unordered_set>
+
+#include "cfg_cleanup_pass.h"
+
+#include "function.h"
+#include "module.h"
+
+namespace spvtools {
+namespace opt {
+
+bool CFGCleanupPass::RemoveUnreachableBlocks(ir::Function* func) {
+  bool modified = false;
+
+  // Mark reachable all blocks reachable from the function's entry block.
+  std::unordered_set<ir::BasicBlock*> reachable_blocks;
+  std::queue<ir::BasicBlock*> worklist;
+  reachable_blocks.insert(func->entry().get());
+  worklist.push(func->entry().get());
+  while (!worklist.empty()) {
+    ir::BasicBlock* block = worklist.front();
+    worklist.pop();
+    block->ForEachSuccessorLabel(
+        [&reachable_blocks, &worklist, this](uint32_t label_id) {
+          auto successor = label2block_[label_id];
+          reachable_blocks.insert(successor);
+          worklist.push(successor);
+        });
+  }
+
+  // Erase unreachable blocks.
+  for (auto ebi = func->begin(); ebi != func->end();)
+    if (reachable_blocks.find(&*ebi) == reachable_blocks.end()) {
+      ebi = ebi.Erase();
```

Comment:
You're too used to LLVM.  Remember that in the current ir::* class hierarchy, instructions only have IDs for the things they use, and the DefUseManager is a separate layer over it.

So deleting an Instruction will not delete the references to that instruction.
Thinks you'll have to clean up:
 - OpName,  possibly OpMemberName (for struct types, so does not apply in this pass)
 - OpDecoration, OpDecorationGroup
 - OpPhi, e.g. you delete A, and so all values defined in A, but that block and its values might be used in an OpPhi in a live block.

Use DefUseManager to trace uses.  It has a convenience KillInst that converts an existing instruction to a Nop and kill the uses.  That'a a common idiom you'll see:  KillInst, then the Module's write-to-output function can be used to skip over all nops.  It works.  However, cleaning up blocks is another matter.   Check Greg's dead-branch-elim code for examples of how to clean up blocks.

And then we have to agree on whether this pass promises to keep up the DefUseManager up to date. By default it does not have to make that promise.  But our direction is to "save time" by agreeing to keep it up to date, and then have the pass manager track whether the DefUseManager is stale or not.